File: src/MCCONS.cpp

#include "../include/MCCONS.h"


struct pairs_sort_predicate
{  // sorting predicate for pairs of size_t, double by double
    bool operator()(const std::pair<int,double> &left,
                    const std::pair<int,double> &right) {
        return left.second < right.second;
    }
};


std::vector< std::vector<RNATree> > strings_to_trees(std::vector< std::vector<std::string> > dot_brackets_vectors)
{   // get the list of trees from the dot bracket strings

    std::vector< std::vector<RNATree> > trees = std::vector< std::vector<RNATree> >();

    for (size_t i = 0; i < dot_brackets_vectors.size(); ++i)
    {
        // initialize a vector of trees to add later
        std::vector<std::string> current_unique_brackets = std::vector<std::string>();
        std::vector<RNATree> current_unique_trees = std::vector<RNATree>();

        for(size_t j = 0; j < dot_brackets_vectors[i].size(); ++j)
        {
            // calculate the bracket
            std::string bracket = only_paired(dot_brackets_vectors[i][j]);

            // check if it has already been processed
            if (std::find(current_unique_brackets.begin(), current_unique_brackets.end(), bracket) == current_unique_brackets.end())
            {
                current_unique_brackets.push_back(bracket);
                current_unique_trees.push_back(RNATree(bracket, '(', ')', '.'));
            }
        }
        // add to the trees
        trees.push_back(current_unique_trees);
    }
    return trees;
}


std::vector< std::vector<std::string> > filter_dot_brackets(std::vector< std::vector<std::string> > dot_brackets_vectors,
                                                            std::vector<std::string> brackets)
{  // remove dot_brackets with the wrong base tree topology
    std::vector< std::vector<std::string> > result = std::vector< std::vector<std::string> >();
    for(unsigned int i = 0; i < dot_brackets_vectors.size(); ++i)
    {
        std::string bracket = brackets[i];
        std::vector<std::string> compliant_dot_brackets = std::vector<std::string>();

        for(unsigned int j = 0; j < dot_brackets_vectors[i].size(); ++j)
        {
            if(only_paired(dot_brackets_vectors[i][j]) == bracket)
            {
                compliant_dot_brackets.push_back(dot_brackets_vectors[i][j]);
            }
        }
        result.push_back(compliant_dot_brackets);
    }
    return result;
}





void MCCONS(std::string path,
            Solver* tree_solver,
            Solver* dot_bracket_solver)
{   // given a solver and a problem instance, apply the MC-Cons algorithm

    // PHASE 1: TREE CONSENSUS

    // data acquisition
    std::vector<std::vector<std::string> > dot_brackets = read_marna_file(path);
    std::vector<std::vector<RNATree> > trees = strings_to_trees(dot_brackets);
    ConsensusProblem<RNATree> tree_problem = ConsensusProblem<RNATree>(trees, *unit_tree_indel_distance_trees);
    int num_comparisons = dot_brackets.size() * (dot_brackets.size() - 1);  // n (n -1)

    // PHASE 1 SOLVING
    bool silent_output = tree_solver->is_silent();
    if (! silent_output)
    {
        // print out the size of the search space
        // long int 
        std::cerr << "Phase 1: Base Pair Tree Consensus (1)" << std::endl;
    }

    std::vector<Solution> tree_consensus = tree_solver->solve(tree_problem.get_distance_matrix(),
                                                              tree_problem.get_ranges());

    // sort the tree consensus by score (lower is better)
    std::vector< std::pair<int, double> > sorted_order = std::vector<std::pair<int, double> >();
    for (size_t i = 0; i != tree_consensus.size(); ++i)
    {
      sorted_order.push_back(std::pair<size_t, double>(i, tree_consensus[i].get_score()));
    }
    std::sort(sorted_order.begin(), sorted_order.end(), pairs_sort_predicate());
    std::vector<Solution> tree_consensus_tmp = std::vector<Solution>();
    for (size_t i = 0; i != sorted_order.size(); ++i)
    {
        tree_consensus_tmp.push_back(tree_consensus[sorted_order[i].first]);
    }
    tree_consensus = tree_consensus_tmp;


    // PHASE 1.5: FILTERING
    std::vector<std::vector< std::vector<std::string> > > prob2_data = std::vector< std::vector< std::vector<std::string> > >();

    // get the brackets
    for(std::vector<Solution>::iterator it = tree_consensus.begin(); it != tree_consensus.end(); ++it) {
        std::vector<std::string> brackets = std::vector<std::string>();
        for(size_t j = 0; j != it->get_genes().size(); ++j) {
            int gene = it->get_genes()[j];
            brackets.push_back(tree_problem.get_objects()[gene].get_brackets());
        }
        // filter and add to the problem 2 data
        std::vector< std::vector<std::string> > filtered_brackets = filter_dot_brackets(dot_brackets, brackets);
        prob2_data.push_back(filtered_brackets);
    }
    std::vector< ConsensusProblem<std::string> > dot_bracket_problems = std::vector< ConsensusProblem<std::string> >();
    // instantiate the problems
    for (size_t i = 0; i != prob2_data.size(); ++i)
    {
        dot_bracket_problems.push_back(ConsensusProblem<std::string>(prob2_data[i], *string_edit_distance));
    }


    // PHASE 2: STRING CONSENSUS
    if (! silent_output)
    {
        // update on the number of optimal trees found
        std::cerr << std::endl << "Phase 2: String Edit Distance on Vienna Dot Bracket (" << dot_bracket_problems.size() << ")" << std::endl;
    }
    // if too many output, make the solver silent and add a progress bar over all tree consensus instead of individually
    bool too_much_progress_bar = (tree_consensus.size() > 10) && (! silent_output);
    if (too_much_progress_bar)
    {
        dot_bracket_solver->flip_silent();
    }
    std::vector< std::vector<Solution> > dot_bracket_consensus = std::vector< std::vector<Solution> >();
    ProgressBar progress = ProgressBar();
    for (size_t i = 0; i != prob2_data.size(); ++i)
    {
        dot_bracket_consensus.push_back(dot_bracket_solver->solve(dot_bracket_problems[i].get_distance_matrix(),
                                                                 dot_bracket_problems[i].get_ranges()));
        progress.update((float) i / tree_consensus.size());
    }
    progress.clean();

    if (! silent_output)
    {
        std::cerr << std::endl;
    }

    // printing the results to stdout
    // print the best dotbracket consensus for each tree consensus
    int gene;
    double tree_score, best_dot_bracket_score;

    int sol_index = 0;
    for(size_t i = 0; i != dot_bracket_consensus.size(); ++i)
    {
        // figure out what the best scores are within each tree consensus
        tree_score = tree_consensus[i].get_score() / num_comparisons;
        best_dot_bracket_score = std::numeric_limits<double>::infinity();
        for(size_t j = 0; j != dot_bracket_consensus[i].size(); ++j)
        {
            if (dot_bracket_consensus[i][j].get_score() < best_dot_bracket_score)
            {
                best_dot_bracket_score = dot_bracket_consensus[i][j].get_score();
            }
        }
        for(size_t j = 0; j != dot_bracket_consensus[i].size(); ++j)
        {
            if (dot_bracket_consensus[i][j].get_score() == best_dot_bracket_score) {
                // output the consensus
                std::cout << "> " << sol_index << " " << tree_score << " " << best_dot_bracket_score / num_comparisons << std::endl;
                sol_index += 1;
                for (size_t gene_index = 0; gene_index != dot_bracket_consensus[i][j].get_genes().size(); ++gene_index) {
                    gene = dot_bracket_consensus[i][j].get_genes()[gene_index];
                    std::cout << dot_bracket_problems[i].get_objects()[gene] << std::endl;
                }
                std::cout << std::endl;
            }
        }
    }

    return;
}

File: java/src/MCCons.java

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

public class MCCons {


    /**
     * get the list of trees from the dot bracket strings
     * @param dotBrackets
     * @return
     */
    public static ArrayList<ArrayList<OrderedRootedLabeledTree>> dotBracketsToTrees(ArrayList<ArrayList<String>> dotBrackets) {

        ArrayList<ArrayList<OrderedRootedLabeledTree>> trees = new ArrayList<>();

        for (ArrayList<String> list : dotBrackets) {
            // initialize a vector of trees to add later
            HashSet<String> uniqueStrings = new HashSet<>();
            ArrayList < OrderedRootedLabeledTree > uniqueTrees = new ArrayList<>();

            for (String string : list) {
                // check if it has already been processed
                if (! uniqueStrings.contains(string))
                {
                    uniqueStrings.add(string);
                    uniqueTrees.add(new OrderedRootedLabeledTree(string, '(', ')', '.'));
                }
            }
            // add to the trees
            trees.add(uniqueTrees);
        }
        return trees;
    }


    /**
     * remove dot_brackets with the wrong base tree topology
     * @param data
     * @param constraints
     * @return
     */
    public static ArrayList<ArrayList<String>> filter_dot_brackets(ArrayList<ArrayList<String>> data,
                                                     ArrayList<String> constraints) {

        ArrayList < ArrayList < String >> filtered = new ArrayList<>();
        for ( int i = 0; i < data.size (); ++i)
        {
            String constraint = constraints.get(i);
            ArrayList <String> fitting = new ArrayList<>();

            for ( String string : data.get(i))
            {
                if (string == constraint) {
                    fitting.add(string);
                }
            }
            filtered.add(fitting);
        }
        return filtered;
    }


    /**
     * given a solver and a problem instance, apply the MC-Cons algorithm
     * @param path path of the input file
     * @param firstSolver solver for the tree problems
     * @param secondSolver solver for the dot bracket problems
     */
    public static void MCCONS(String path,
                              Solver firstSolver,
                              Solver secondSolver) throws IOException{

        // PHASE 1: TREE CONSENSUS
        // data acquisition
        ArrayList < ArrayList < String >>dot_brackets = Readers.readMarnaFile(path);
        ArrayList < ArrayList < OrderedRootedLabeledTree >>trees = dotBracketsToTrees(dot_brackets);
        ConsensusProblem<OrderedRootedLabeledTree> treeProblem = new ConsensusProblem <> (trees, new TreeEditDistance());
        int num_comparisons = dot_brackets.size() * (dot_brackets.size() - 1);  // n (n -1)
        // PHASE 1 SOLVING
        boolean verbose = firstSolver.isVerbose();
        if (verbose) {
            System.err.println("Phase 1: Base Pair Tree Consensus (1)");
        }

        ArrayList < Solution > treeConsensus = firstSolver.solve(treeProblem.getDistanceMatrix(),
                                                                  treeProblem.getRanges());

        // sort the tree consensus by score (lower is better)
        ScoreComparator<Integer> comparator = new ScoreComparator<>();
        ArrayList < Pair <Integer, Double>> sortedOrder = new  ArrayList <>();
        for (int i = 0; i != treeConsensus.size(); ++i) {
            sortedOrder.add(new Pair<>(i, treeConsensus.get(i).getScore()));
        }
        Collections.sort(sortedOrder, comparator);

        ArrayList <Solution> tree_consensus_tmp = new ArrayList<> ();
        for (int i = 0; i != sortedOrder.size(); ++i) {
            tree_consensus_tmp.add(treeConsensus.get(sortedOrder.get(i).getFirst()));
        }
        treeConsensus = tree_consensus_tmp;


        // PHASE 1.5: FILTERING
        ArrayList < ArrayList < ArrayList < String >>> prob2_data = new ArrayList<>();

        // get the brackets
        for (Solution solution: treeConsensus)
        {
            ArrayList < String > brackets = new ArrayList<>();
            for (Integer gene : solution.getGenes()){
                    brackets.add(treeProblem.getObjects().get(gene).toString());
            }

            // filter and add to the problem 2 data
            ArrayList < ArrayList < String >>filtered_brackets = filter_dot_brackets(dot_brackets, brackets);
            prob2_data.add(filtered_brackets);
        }
        ArrayList < ConsensusProblem < String >> dot_bracket_problems = new ArrayList <>();

        // instantiate the problems
        for (int i = 0; i != prob2_data.size(); ++i) {
            dot_bracket_problems.add(new ConsensusProblem<>(prob2_data.get(i), new StringEditDistance()));
        }


        // PHASE 2: STRING CONSENSUS
        if (verbose) {
            // update on the number of optimal trees found
            System.err.println(System.lineSeparator() + "Phase 2: String Edit Distance on Vienna Dot Bracket (" +
                    dot_bracket_problems.size() + ")");
        }
        // if too many output, make the solver silent and add a progress
        // bar over all tree consensus instead of individually
        if ((treeConsensus.size() > 10) && (verbose)) {
            verbose = (!verbose);
        }
        ArrayList < ArrayList < Solution >>dotBracketConsensus = new ArrayList <>();
        ProgressBar progress = new ProgressBar("", 40);
        for (int i = 0; i != prob2_data.size(); ++i) {
            dotBracketConsensus.add(secondSolver.solve(dot_bracket_problems.get(i).getDistanceMatrix(),
                    dot_bracket_problems.get(i).getRanges()));
            progress.update((float) i / treeConsensus.size());
        }
        progress.clean();

        if (verbose) {
            System.err.println();
        }

        // printing the results to stdout
        // print the best dotbracket consensus for each tree consensus
        double tree_score, best_dot_bracket_score;
        int sol_index = 0;
        int treeProblemIndex = 0;
        for (ArrayList<Solution> solutions : dotBracketConsensus){
            // figure out what the best scores are within each tree consensus
            tree_score = treeConsensus.get(treeProblemIndex).getScore() / num_comparisons;
            best_dot_bracket_score = Double.POSITIVE_INFINITY;

            for (Solution solution : solutions)
            {
                if (solution.getScore() < best_dot_bracket_score)
                {
                    best_dot_bracket_score = solution.getScore();
                }
            }

            for (Solution solution : solutions)
            {
                if (solution.getScore() == best_dot_bracket_score)
                {
                    // output it
                    double normalizedScore = best_dot_bracket_score / num_comparisons;
                    StringBuilder builder = new StringBuilder();
                    builder.append("> ");
                    builder.append(sol_index);       // index
                    builder.append(" ");
                    builder.append(tree_score);      // first score
                    builder.append(" ");
                    builder.append(normalizedScore); // normalized second score
                    builder.append(System.lineSeparator());
                    for (Integer gene : solution.getGenes()){
                        String object = dot_bracket_problems.get(treeProblemIndex).getObjects().get(gene);
                        builder.append(object);
                        builder.append(System.lineSeparator());
                    }
                    System.out.println(builder.toString());
                    sol_index += 1;
                }
            }
            treeProblemIndex+=1;
        }
    }
}